atintegrators/SimpleRadiationPass.c

```diff
@@ -0,0 +1,148 @@
+
+#include "atelem.c"
+#include "atrandom.c"
+
+struct elem 
+{
+  double *damp_mat_diag;
+  double dispx;
+  double dispxp;
+  double dispy;
+  double dispyp;
+  double U0;
+  double EnergyLossFactor;
+};
+
+void SimpleRadiationPass(double *r_in,
+           double *damp_mat_diag, double dispx, double dispxp,
+           double dispy, double dispyp, double EnergyLossFactor, int num_particles)
+
+{
+  double *r6;
+  int c, i;
+  double x,xp,y,yp,z,dpp;
+  
+  #pragma omp parallel for if (num_particles > OMP_PARTICLE_THRESHOLD*10) default(shared) shared(r_in,num_particles) private(c,r6)
+  for (c = 0; c<num_particles; c++) { /*Loop over particles  */
+    r6 = r_in+c*6;
+    
+    if(!atIsNaN(r6[0])) {
+
+      dpp = r6[4];
+      x = r6[0] - dispx*dpp;
+      xp = r6[1] - dispxp*dpp;
+      
+      y = r6[2] - dispy*dpp;
+      yp = r6[3] - dispyp*dpp;
+      
+      z = r6[5];        
+    
+      if(damp_mat_diag[0]!=1.0) {
+        x *= damp_mat_diag[0];
+      }
+
+      if(damp_mat_diag[1]!=1.0) {
+        xp *= damp_mat_diag[1];
+      }
+      
+      if(damp_mat_diag[2]!=1.0) {
+        y *= damp_mat_diag[2];
+      }
+
+      if(damp_mat_diag[3]!=1.0) {
+        yp *= damp_mat_diag[3];
+      }
+      
+      if(damp_mat_diag[4]!=1.0) {
+        dpp *= damp_mat_diag[4];
+      }
+
+      if(damp_mat_diag[5]!=1.0) {
+        z *= damp_mat_diag[5];
+      }
+
+      r6[4] = dpp;
+      r6[5] = z;
+      r6[0] = x + dispx*dpp;
+      r6[1] = xp + dispxp*dpp;
+      r6[2] = y + dispy*dpp;
+      r6[3] = yp + dispyp*dpp;      
+      
+      r6[4] -= EnergyLossFactor;      
+    }
+  }
+}
+
+#if defined(MATLAB_MEX_FILE) || defined(PYAT)
+ExportMode struct elem *trackFunction(const atElem *ElemData,struct elem *Elem,
+                double *r_in, int num_particles, struct parameters *Param)
+{
+/*  if (ElemData) {*/
+        if (!Elem) {
+            double U0, EnergyLossFactor;
+            double dispx, dispxp, dispy, dispyp;
+            double *damp_mat_diag;
+            
+            damp_mat_diag=atGetDoubleArray(ElemData,"damp_mat_diag"); check_error();
+            U0=atGetDouble(ElemData,"U0"); check_error();
+            dispx=atGetDouble(ElemData,"dispx"); check_error();
+            dispxp=atGetDouble(ElemData,"dispxp"); check_error();
+            dispy=atGetDouble(ElemData,"dispy"); check_error();
+            dispyp=atGetDouble(ElemData,"dispyp"); check_error();

```

Comment:
Dispersions are not in the arguments of the `SimpleRadiation` constructor: it's possible to create an element without them. The `simple_ring` utility puts them, but someone could create a `SImpleElement` directly. So it would be safer to declare the dispersions here as optional with default value 0:
```C
dispx=atGetOptionalDouble(ElemData,"dispx", 0.0); check_error();
```